### ISSUE TITLE
Some issue about validation

### ISSUE THREAD

**Issue by hwc1824** [wentaozhu/DeepLung#105]
Hi, Wentao
I'm trying to run your code and I got some issues.
I've followed the instructions in readme and finish 10 fold of detector training.
After that, I try to run frocwrtdetpepchluna16.py to get validation result of each epochs but I got some problem here.
And then I check previous issue to get some clue and still kinda confused.
This is what i think I should do:

1. After training, I got 150 ckpt and val result (pbb, lbb) for each fold
2. Generate the csv file for each validation result for each fold by running frocwrtdetpepchluna16.py
3. Pick best from 150 validation result for each fold and concatenate the csv (`predanno-1d3.csv` or `predanno-1.5d3.csv` I guess)
4. run noduleCADEvaluationLUNA16.py with concatenated prediction csv to get final result

---

And now, I've done the step1 and been able to run frocwrtdetpepchluna16.py
I know that I should uncomment line 144 to get the prediction csv for each epoch in one fold
and I can run the frocwrtdetpepchluna16.py with annotation csv provided by LUNA Like this
```
fold = 1
annotations_filename = "/data/home/hwc/md/luna/DeepLung/evaluationScript/annotations/annotations.csv"# path for ground truth annotations for the fold
annotations_excluded_filename = "/data/home/hwc/md/luna/DeepLung/evaluationScript/annotations/annotations_excluded.csv"# path for excluded annotations for the fold
seriesuids_filename = "/data/home/hwc/md/luna/DeepLung/evaluationScript/annotations/seriesuids.csv"# path for seriesuid for the fold
results_path = "/data/home/hwc/md/luna/DeepLung/detector/results/res18/retrft96"+str(fold)+"/val"#val' #val' ft96'+'/val'#
sideinfopath = '/data/home/hwc/md/luna/data/processed/subset'+str(fold)+'/' # +str(fold)
datapath = '/data/home/hwc/md/luna/data/raw/subset'+str(fold)+'/'
```
but I noticed your reply in https://github.com/wentaozhu/DeepLung/issues/42#issuecomment-435445688 and ,You mentioned that 

> For each fold, you need to use the fold's annotation to run the frocwrtdetpepchluna16.py

**I think that's where I get stucked, where do i get the `fold's annotation` ?**

---

I write a script to split to original annotation.csv into 10 sub_annotations.csv.
each sub_annotations.csv only contains nodule annotation of 1 fold.
then I fill it into `annotations_filename` and this is what I get from fold 1,
```
...
/data/home/hwc/md/luna/DeepLung/detector/results/res18/retrft961/val75/predanno-1.5d3.csv
/data/home/hwc/md/luna/DeepLung/detector/results/res18/retrft961/val75/predanno-1d3.csv  
[0.8973214285714286, 0.890625]                                                           
ep 75 detp -1.5 0.897321428571                                                           
ep 75 detp -1 0.890625                                                                   
/data/home/hwc/md/luna/DeepLung/detector/results/res18/retrft961/val76/predanno-1.5d3.csv
/data/home/hwc/md/luna/DeepLung/detector/results/res18/retrft961/val76/predanno-1d3.csv  
[0.8828125, 0.8828125]                                                                   
ep 76 detp -1.5 0.8828125                                                                
ep 76 detp -1 0.8828125                                                                  
...                                             
```
does this mean that it run correctly?

**Comment by wentaozhu** [comment]
Yes. That is how to create the annotation of each fold. That (ground truth) is used to calculate the FROC.

**Comment by hwc1824** [comment]
Thanks for your fast reply.
I got couple questions about the output of frocwrtdetpepchluna16.py

1. I nocite that in the `evaluationScript`, there is two similar file `noduleCADEvaluationLUNA16.py` and `noduleCADEvaluationLUNA16compare.py`. I know that the former is official released and the latter is the one you revised. When I run the `frocwrtdetpepchluna16.py` script, I have to import latter instead of former (which is the original one) to get it work. Is that correct? 

2. I notice that `detp = [-1.5, -1]` in the script, which i believe is the probability threshold to the detection. If I recall correctly, you set it as `-2` in your paper. does this mean that -1 or -1.5 may yields better performance?

2.  After running `frocwrtdetpepchluna16.py`,  i got something like this ```1 0.930803571429 4 0 ```
I know that the meaning of `fold, FROC, num of epoch - 1, detp - 1`
which means that in fold 1, use dep=-1 and epoch=5 to get froc=0.930803571429
what the meaning of the froc here? AUC of FROC?

Again, thank you for the fast reply,

**Comment by wentaozhu** [comment]
1 I forget. You may look into the code and can easily find it.
2 Yes. You may try different threshold as detection probability.
3 Yes. You may check our paper to get to know the FROC. https://arxiv.org/abs/1801.09555

Please add our papers into your references if you find the repository is helpful to you! Thank you so much for your interest!

Deeplung: 3d deep convolutional nets for automated pulmonary nodule detection and classification, WACV 2018
DeepEM: Deep 3D ConvNets With EM For Weakly Supervised Pulmonary Nodule Detection, MICCAI 2018

**Comment by MjdMahasneh** [comment]
> 3\. Pick best from 150 validation result for each fold and concatenate the csv

Hello @hwcasus, hope you are well. Could you kindly elaborate on this, why do we need to pick the 150 best results? and do you mean pick 150 from each fold, or from each scan?

Looking forward to hearing back from you :)

**Comment by wentaozhu** [comment]
pick 150 for each fold. It is standard cross validation.

**Comment by MjdMahasneh** [comment]
Hello @wentaozhu, I hope you're well. and thank you for you answer, I am aware of that, I might have misunderstood the comment and thought that it was referring to picking 150 predictions.

Maybe you can advise me on my issue, I am running the training script using Python 3.5 (and LUNA-16 dataset) and have been trying to reproduce the results from the paper, for some reason, the model is producing a very high number of bounding boxes predictions per scan (~1 million or more sometimes). Have you come across this issue before? I am not sure why this is happening. Obviously, this also is slowing the evaluation drastically (non-maximum suppression becomes extremely slow and takes hours, I couldn't finish it testing because of that.).

Could this be the difference in python versions (although I have checked the code for inconsistencies between python 2 and 3, namely, I have checked all division lines in all scripts to make sure they are consistent).

Another thing I have noticed is the use of testthresh=-3 in train_detector_se.py, which I can't quite understand it clearly, it seems to me that it's being used for producing a mask...!

I have been working on this for a good while now, so I would really appreciate any advice or assistance :) Many thanks in advance :)
